### PR TITLE
reside-123: Basic file operations

### PR DESCRIPTION
**Pull request by richfitz**
This PR introduces basic file operations to pointr, having reverse-engineered a bit what [shareplum](https://github.com/jasonrollins/shareplum) does. Upload was a particular pain, but most of that was me being daft.

The basic idea is to have a "folder" object and basic operations that apply to a folder (listing, download, upload, go up the tree, go down the tree).  This should make workflows like "download a bunch of files" or "upload a bunch of files" relatively easy to implement.

Not done here is deletion and folder creation

### CODE REVIEW

File: R/sharepoint_client.R [hunk 1]
```diff
@@ -76,6 +76,18 @@ sharepoint_client <- R6::R6Class(
     request = function(verb, path, ...) {
       url <- paste(self$sharepoint_url, path, sep = "/")
       verb(url, ..., handle = private$handle)
+    },
+
+    #' @description
+    #' Get Sharepoint's "Request Digest" security feature.  This method
```
Comment (r-ash) [inline R/sharepoint_client.R:82]: Nice - I read about this before but didn't really understand what it was used for.

File: R/sharepoint_folder.R [hunk 2]
```diff
@@ -0,0 +1,130 @@
...
+    #' @param progress Display httr's progress bar?
+    download = function(path, dest = NULL, progress = FALSE) {
+      url <- sprintf("%s/Files('%s')/$value",
+                     private$api_root, URLencode(path))
+      dest <- dest %||% tempfile_inherit_ext(path)
+      path_show <- sprintf("%s:%s/%s", private$site, private$path, path)
+      download(private$client, url, dest, path_show, progress)
+    },
+
+    #' @description Upload a fole into a folder
```
Comment (r-ash) [inline R/sharepoint_folder.R]: "Upload a file into a folder"

File: R/sharepoint_folder.R [hunk 3]
```diff
@@ -0,0 +1,130 @@
...
+  public = list(
+    initialize = function(client, site, path, verify = FALSE) {
+      private$client <- client
+      private$site <- site
+      private$path <- path
+      private$api_root <- sprintf(
+        "/sites/%s/_api/web/GetFolderByServerRelativeURL('%s')",
+        site, URLencode(path))
+
+      if (verify) {
```
Comment (r-ash) [inline R/sharepoint_folder.R:26]: This currently not working with the right client. And testing not showing this issue.

Getting 
```
 Error in .subset2(public_bind_env, "initialize")(...) : 
  Internal Server Error (HTTP 500). 
```

Whenever `verify = TRUE`

**Review by r-ash — CHANGES_REQUESTED** [review]
This looks great - I tested this locally. One issue with when `verify = TRUE` which we've discussed already.

Also noticed that some of the relative paths aren't quite working I don't think.

e.g. doing
```
meetings_folder$download("Previous presentations/Meeting slide ups.pptx", t)
```
Fails to locate the file with message 
```
Error in download(private$client, url, dest, path_show, progress) : 
  Remote file not found at 'fom/sph/hiv-inference-group:docs/Monday meeting slides/Previous presentations/Meeting slide ups.pptx'
```

but doing it bit by bit succeeds
```
prev <- meetings_folder$folder("Previous presentations")
prev$download("Meeting slide ups.pptx", t)
```
It would also be cool if we could have an option to print file sizes in human readable format. Thinking like `-h` effect in terminal. But that's fairly minor.

File: R/sharepoint_folder.R (hunk 3 above)
Comment (richfitz) [reply R/sharepoint_folder.R:26]: This is the right client, and works for me if constructed using public functions.  I will add an assertion in though to stop the confusing way

**Review by r-ash — APPROVED** [review]
Nice one

**Pull request merged**